core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1018,34 +1033,48 @@ private IsReadyForExecutionResult isReadyForExecution(SubPlan subPlan)
                     return IsReadyForExecutionResult.notReady();
                 }
 
-                Optional<OutputDataSizeEstimateResult> result = sourceStageExecution.getOutputDataSize(stageExecutions::get);
+                Optional<OutputDataSizeEstimateResult> result = sourceStageExecution.getOutputDataSize(stageExecutions::get, eager);
                 if (result.isEmpty()) {
                     return IsReadyForExecutionResult.notReady();
                 }
 
                 switch (result.orElseThrow().getStatus()) {
                     case ESTIMATED_BY_PROGRESS -> estimatedByProgressSourcesCount++;
                     case ESTIMATED_BY_SMALL_INPUT -> estimatedBySmallInputSourcesCount++;
+                    case ESTIMATED_FOR_EAGER_PARENT -> estimatedForEagerParent++;
                     default -> throw new IllegalStateException(format("unexpected status %s", result.orElseThrow().getStatus())); // FINISHED handled above
                 }
 
                 sourceOutputSizeEstimates.put(sourceStageExecution.getStageId(), result.orElseThrow().getOutputDataSizeEstimate());
                 someSourcesMadeProgress = someSourcesMadeProgress || sourceStageExecution.isSomeProgressMade();
             }
 
-            if (!subPlan.getChildren().isEmpty() && !someSourcesMadeProgress) {
+            if (!subPlan.getChildren().isEmpty() && !someSourcesMadeProgress && !eager) {
                 return IsReadyForExecutionResult.notReady();
             }
 
             if (speculative) {
-                log.debug("scheduling speculative %s/%s; sources: finished=%s; estimatedByProgress=%s; estimatedSmall=%s",
+                log.debug("scheduling speculative %s/%s; sources: finished=%s; estimatedByProgress=%s; estimatedSmall=%s; estimatedForEagerParent=%s",
                         queryStateMachine.getQueryId(),
                         subPlan.getFragment().getId(),
                         finishedSourcesCount,
                         estimatedByProgressSourcesCount,
-                        estimatedBySmallInputSourcesCount);
+                        estimatedBySmallInputSourcesCount,
+                        estimatedForEagerParent);
             }
-            return IsReadyForExecutionResult.ready(sourceOutputSizeEstimates.buildOrThrow(), false);
+            return IsReadyForExecutionResult.ready(sourceOutputSizeEstimates.buildOrThrow(), eager);
+        }
+
+        private boolean shouldScheduleEagerly(SubPlan subPlan)
+        {
+            return hasSmallFinalLimitNode(subPlan);
+        }
+
+        private static boolean hasSmallFinalLimitNode(SubPlan subPlan)
+        {
+            return PlanNodeSearcher.searchFrom(subPlan.getFragment().getRoot())
```

Comment:
Is it possible for a plan to be more complex then a single `LIMIT` node? (For example `Final Aggregation` -> `Limit`).

What do you think about having an extra condition if the stage is distribution "SINGLE"?

Reply:
> Is it possible for a plan to be more complex then a single LIMIT node? (For example Final Aggregation -> Limit).

I think it is possible - and it would still be caught by this predicate as we are just looking for final `Limit` node in fragment - not assuming it is root.

> What do you think about having an extra condition if the stage is distribution "SINGLE"?

As part of predicate or more like an assertion? I think we should always get "SINGLE" distribution for fragment with final Limit node, right?

Reply:
Changed method to:
```java
        private static boolean hasSmallFinalLimitNode(SubPlan subPlan)
        {
            if (subPlan.getFragment().getPartitioning().isSingleNode()) {
                // Final LIMIT should always have SINGLE distribution
                return false;
            }
            return PlanNodeSearcher.searchFrom(subPlan.getFragment().getRoot())
                    .where(node -> node instanceof LimitNode limitNode && !limitNode.isPartial() && limitNode.getCount() < 1_000_000)
                    .matches();
        }

```